app/src/main/kotlin/com/wire/android/GlobalObserversManager.kt

```diff
@@ -91,4 +100,20 @@ class GlobalObserversManager @Inject constructor(
                 // but we can't start PersistentWebSocketService here, to avoid ForegroundServiceStartNotAllowedException
             }
     }
+
+    private fun CoroutineScope.handleLogouts() {
+        callbackFlow<Unit> {
+            val callback: LogoutCallback = object : LogoutCallback {
+                override suspend fun invoke(userId: UserId, reason: LogoutReason) {
+                    notificationManager.stopObservingOnLogout(userId)
+                    notificationChannelsManager.deleteChannelGroup(userId)
+                    if (reason != LogoutReason.SELF_SOFT_LOGOUT) {
+                        userDataStoreProvider.getOrCreate(userId).clear()
+                    }
+                }
+            }
+            coreLogic.getGlobalScope().logoutCallbackManager.register(callback)
+            awaitClose { coreLogic.getGlobalScope().logoutCallbackManager.unregister(callback) }
+        }.launchIn(this)
+    }
```

Comment:
Initially I did it like that - with a `SharedFlow`, but this way we're making it so that the callback is executed independently and logout doesn't wait for it to complete. Unfortunately, [`SharedFlow` doesn't wait for a subscriber to complete collecting](https://github.com/Kotlin/kotlinx.coroutines/issues/2603), documentation is misleading. It's probably possible with single thread dispatcher but then `logout` and `observe` would need to be executed on that same single thread and it feels already more tricky. Current solution calls a suspend function directly so it's up to the app to decide whether it wants to execute it sequentially as a part of logout or just launch another coroutine to do that. There actually was one issue with that, when resetting app lock and logging out all accounts, that it required to make all logout actions before the `LogoutUseCase` returns a result. I would prefer to keep it this way, but I'm also curious to know if you have any other ideas on this. 😉

Reply:
> Initially I did it like that - with a `SharedFlow`, but this way we're making it so that the callback is executed independently and logout doesn't wait for it to complete. Unfortunately, [`SharedFlow` doesn't wait for a subscriber to complete collecting](https://github.com/Kotlin/kotlinx.coroutines/issues/2603), documentation is misleading. It's probably possible with single thread dispatcher but then `logout` and `observe` would need to be executed on that same single thread and it feels already more tricky. Current solution calls a suspend function directly so it's up to the app to decide whether it wants to execute it sequentially as a part of logout or just launch another coroutine to do that. There actually was one issue with that, when resetting app lock and logging out all accounts, that it required to make all logout actions before the `LogoutUseCase` returns a result. I would prefer to keep it this way, but I'm also curious to know if you have any other ideas on this. 😉

Ooh, great find. I absolutely didn't know that. Thought it would wait for completion. You definitely thought it all through 😍